src/main/scala/com/gu/invoicing/invoice/Model.scala

```diff
@@ -0,0 +1,274 @@
+package com.gu.invoicing.invoice
+
+import java.time.format.DateTimeFormatter
+import java.time.format.DateTimeFormatter.ofPattern
+import java.time.{LocalDate, LocalDateTime}
+import java.util.UUID
+import com.gu.spy._
+
+/**
+ * Needed for upickle to handle optional fields
+ * http://www.lihaoyi.com/upickle/#CustomConfiguration
+ */
+class OptionPickler extends upickle.AttributeTagged {
+  implicit def optionWriter[T: Writer]: Writer[Option[T]] =
+    implicitly[Writer[T]].comap[Option[T]] {
+      case None => null.asInstanceOf[T]
+      case Some(x) => x
+    }
+
+  implicit def optionReader[T: Reader]: Reader[Option[T]] = {
+    new Reader.Delegate[Any, Option[T]](implicitly[Reader[T]].map(Some(_))){
+      override def visitNull(index: Int) = None
+    }
+  }
+}
+
+/**
+ * Data models and JSON codecs
+ */
+object Model extends OptionPickler {
+  case class Config(clientId: String, clientSecret: String)
+  case class AccessToken(access_token: String)
+
+  case class Invoices(
+    invoices: List[Invoice],
+    success: Boolean
+  )
+  case class Invoice(
+    id: String,
+    accountId: String,
+    accountNumber: String,
+    accountName: String,
+    invoiceDate: LocalDate,
+    invoiceNumber: String,
+    dueDate: LocalDate,
+    invoiceTargetDate: LocalDate,
+    amount: BigDecimal,
+    balance: BigDecimal,
+    creditBalanceAdjustmentAmount: BigDecimal,
+    createdBy: String,
+    status: String,
+    body: String,
+    invoiceItems: List[InvoiceItem],
+    invoiceFiles: List[InvoiceFile],
+  )
+  case class InvoiceItem(
+    id: String,
+    subscriptionName: String,
+    subscriptionId: String,
+    serviceStartDate: LocalDate,
+    serviceEndDate: LocalDate,
+    chargeAmount: BigDecimal,
+    chargeDescription: String,
+    chargeName: String,
+    chargeId: String,
+    productName: String,
+    quantity: BigDecimal,
+    taxAmount: BigDecimal,
+    unitOfMeasure: String,
+    chargeDate: String, // FIXME this has different format from other kinds of localdatetimes
+    chargeType: String,
+    processingType: String,
+    appliedToItemId: Option[String]
+  )
+  case class InvoiceFile(
+    id: String,
+    versionNumber: Int,
+    pdfFileUrl: String
+  )
+
+  case class InvoiceWithPayment(
+    subscriptionName: String,
+    date: LocalDate,
+    paymentMethod: PaymentMethod,
+    price: Double,
+    downloadUrl: String,
```

Comment:
In that this isn't a full URL or even a relative URL, I wonder if it should be something like `downloadUrlPart` or `downloadIdentifier`, so the URL can be built later in the client?

Alternatively in anticipation of the download endpoint existing, could it generate the full URL that the client can blindly use (a practice I've observed in various REST APIs). Would almost definitely need some work in `manage-frontend` to pass the `Origin`/`Host` header or something so it could build a valid manage accessible URL (i.e. a manage URL).

Reply:
https://github.com/guardian/invoicing-api/pull/7/commits/7cf1bb4eb26b9263dbb10e727e4dd56633fea960 - Nice 👍